include/crab/domains/object_domain.hpp

```diff
@@ -0,0 +1,3681 @@
+  /// @param rgn a reference variable or an object id
+  /// @return if we find the ghost, returns; otherwise, abort with an error
+  variable_t get_base_addr_or_fail(const variable_t &v) const {
+    assert(v.get_type().is_reference() || v.get_type().is_region());
+    if (!m_refs_base_addrs_map) {
+      CRAB_ERROR(domain_name(), "::", __func__,
+                 ": base address map is not created");
+    }
+    auto it = (*m_refs_base_addrs_map).find(v);
+    if (it == (*m_refs_base_addrs_map).end()) {
+      CRAB_ERROR(domain_name(), "::", __func__, ": ", v, " not found");
+    }
+    return it->second;
+  }
+
+  /// @brief check whether two reference variable refer the same memory object
+  /// @param x a reference variable
+  /// @param y a reference variable
+  /// @return true if x_base == y_base satisfies in address domain
+  bool test_two_addrs_equality(const variable_t &x, const variable_t &y) {
+    return m_addrs_dom.equals(get_or_insert_base_addr(x),
+                              get_or_insert_base_addr(y));
+  }
+
+  /// @brief check the reference refer the MRU object
+  /// @param ref a reference variable
+  /// @param id an object id
+  /// @return true if ref_base == id_mru_base satisfies in address domain
+  bool test_ref_refer_mru_object(const variable_t &ref, const obj_id_t &id) {
+    return test_two_addrs_equality(ref, id);
+  }
+
+  /***************** Reg-fld domain operations *****************/
+  /// @brief get or create a symbolic variable in fld-reg equality domain
+  /// @param eq_dom an equality domain value
+  /// @param variable a field
+  /// @return a symbolic variable such as #var1
+  usymb_t get_or_insert_symbol(eq_fields_abstract_domain_t &eq_dom,
+                               const flds_dom_variable_t &variable) {
+    std::shared_ptr<usymb_t> symb_ptr = eq_dom.get(variable);
+    if (symb_ptr) {
+      return *symb_ptr;
+    } else {
+      return symbolic_variable_equiality_domain_impl::make_fresh_var_symbol<
+          symbolic_variable_equiality_domain_impl::symbolic_var>();
+    }
+  }
+
+  /***************** ODI map operations *****************/
+  /// @brief move singleton object into odi map
+  /// @pre the precondition is the object now becomes non-singleton
+  /// @param id an object id
+  void move_singleton_to_odi_map(const obj_id_t &id) {
+    crab::CrabStats::count(domain_name() + ".count.move_object");
+    crab::ScopedCrabStats __st__(domain_name() + ".move_object");
+
+    base_dom_variable_vector_t flds_vec;
+    get_obj_flds_with_ghost(id, flds_vec);
+
+    const odi_domain_product_t *prod_ref = m_odi_map.find(id);
+    if (!prod_ref) {
+      CRAB_ERROR(domain_name(), "::", __func__, ": object ", id,
+                 " value is not found on odi map");
+    }
+    // NOTE: copy the following value is required because
+    // updating the odi map is construcing a new tree based on previous tree
+    odi_domain_product_t res_prod = *prod_ref;
```

Comment:
Yes, even if a singleton object is in the base because the ODI map stores a pair of `<object id, <info, value>>` where the info stores cache status and reference counting. If a singleton in base, the `value` is a top. Copying a top value is cheap.